### PR TITLE
Shim vote extensions for Tendermint 0.37.x

### PR DESCRIPTION
**Pull request by sug0**
Implements the fifth item of issue https://github.com/anoma/namada/issues/242

With the `abcipp` feature flag enabled, test this PR with this Tendermint version:
https://github.com/heliaxdev/tendermint/releases/tag/v0.1.1-abcipp

Otherwise, use this version:
https://github.com/tendermint/tendermint/tree/0b58342a46b1917a40d506e48e87448dcc8b3131

---

**Note:** Either websockets rpc code, or the pubsub layer in Tendermint are broken, so client transfers performed with `namadac` may hang indefinitely, and e2e tests are definitely broken because of them.

### CODE REVIEW

**Comment by sug0** [comment]
pls update wasm

**Comment by sug0** [comment]
pls update wasm

File: apps/src/lib/client/tx.rs [hunk 1]
```diff
@@ -1195,56 +1188,80 @@ pub async fn broadcast_tx(
...
-    let wrapper_query = Query::from(EventType::NewBlockHeader)
-        .and_eq(ACCEPTED_QUERY_KEY, wrapper_hash.as_str());
-    let tx_query = Query::from(EventType::NewBlockHeader)
-        .and_eq(APPLIED_QUERY_KEY, decrypted_hash.as_str());
+    // It is better to subscribe to the transaction before it is broadcast
+    //
+    // Note that the `APPLIED_QUERY_KEY` key comes from a custom event
+    // created by the shell
+    let query = Query::from(EventType::NewBlock)
+        .and_eq(APPLIED_QUERY_KEY, wrapper_hash.as_str());
```
Comment (yito88) [inline apps/src/lib/client/tx.rs]:
```suggestion
        .and_eq(ACCEPTED_QUERY_KEY, wrapper_hash.as_str());
```

File: apps/src/lib/client/tx.rs [hunk 2]
```diff
@@ -1195,56 +1188,80 @@ pub async fn broadcast_tx(
...
+
+    // We also subscribe to the event emitted when the encrypted
+    // payload makes its way onto the blockchain
+    let mut decrypted_tx_subscription = {
+        let mut decrypted_tx_subscription = TendermintWebsocketClient::open(
+            WebSocketAddress::try_from(address.clone())?,
+            None,
+        )?;
+        let query = Query::from(EventType::NewBlock)
+            .and_eq(ACCEPTED_QUERY_KEY, decrypted_hash.as_str());
```
Comment (yito88) [inline apps/src/lib/client/tx.rs]:
```suggestion
        .and_eq(APPLIED_QUERY_KEY, wrapper_hash.as_str());
```

File: apps/src/lib/client/tendermint_rpc_types.rs [hunk 3]
```diff
@@ -55,6 +39,54 @@ pub struct TxResponse {
...
+    /// Searches for custom events emitted from the ledger and converts
+    /// them back to thin wrapper around a hashmap for further parsing.
+    pub fn parse(
+        json: serde_json::Value,
+        event_type: NamadaEventType,
+        tx_hash: &str,
+    ) -> Result<Option<TxResponse>, Error> {
+        let mut selector = jsonpath::selector(&json);
+        let mut event = {
+            match selector(&format!("$.events.[?(@.type=='{}')]", event_type))
```
Comment (yito88) [inline apps/src/lib/client/tendermint_rpc_types.rs]:
```suggestion
            match selector(&format!(
                "$.data.value.result_end_block.events[?(@.type == '{}')]",
                event_type
            ))
```

**Review by batconjurer — APPROVED** [review]

**Comment by james-chf** [comment]
Do we still use the Tendermint version https://github.com/heliaxdev/tendermint/releases/tag/v0.1.1-abcipp when running the ledger with the changes from this branch?

**Comment by sug0** [comment]
> Do we still use the Tendermint version https://github.com/heliaxdev/tendermint/releases/tag/v0.1.1-abcipp when running the ledger with the changes from this branch?

Depends on whether you have `abcipp` enabled or not. If not, you run [this tendermint](https://github.com/tendermint/tendermint/tree/0b58342a46b1917a40d506e48e87448dcc8b3131)

**Comment by james-chf** [comment]
> > Do we still use the Tendermint version https://github.com/heliaxdev/tendermint/releases/tag/v0.1.1-abcipp when running the ledger with the changes from this branch?
> 
> Depends on whether you have `abcipp` enabled or not. If not, you run [this tendermint](https://github.com/tendermint/tendermint/tree/0b58342a46b1917a40d506e48e87448dcc8b3131)

We should specify this in the repo somewhere, especially since it is the default Tendermint required (i.e. the default ledger build won't run without it). There isn't a release yet at https://github.com/heliaxdev/tendermint/releases but we should make one, then we can update `get_tendermint.sh`

File: vm_env/Cargo.toml [hunk 4]
```diff
@@ -7,10 +7,18 @@ resolver = "2"
...
-default = []
+default = ["abciplus"]
+
+abciplus = [
+  "namada/abciplus",
+]
+
+abcipp = [
+  "namada/abcipp",
+]
```
Comment (james-chf) [inline vm_env/Cargo.toml:18]: I think we should use something other than `abciplus`/`abcipp` for the feature flag names that is more standard. Maybe `tendermint0.37` and `tendermint0.38` respectively, or `abci0.18` and `abci0.19` (although I'm not sure if that new ABCI semver terminology is official yet)

**Comment by sug0** [comment]
pls update wasm

**Review by james-chf — APPROVED** [review]

**Comment by sug0** [comment]
pls update wasm

**Pull request merged**